File: ts/src/audit.ts
import { readdirSync, statSync } from "node:fs"
import { resolve, relative } from "node:path"

const FILE_REF_PATTERN = /(?:^|\s)(?:`)?((?:[\w.-]+\/)*[\w.-]+\.\w+)(?:`)?(?=\s|$|[.,;!?])/gm
const EXTRA_IGNORE = new Set([".git", "node_modules", "__pycache__", ".pytest_cache", ".ruff_cache", ".mypy_cache"])

export interface NearMatch {
  claimed: string
  found: string
  similarity: number
}

export interface AuditResult {
  prompt: string
  claimedFiles: string[]
  actualFiles: string[]
  missingFiles: string[]
  extraFiles: string[]
  matchedFiles: string[]
  nearMatches: NearMatch[]
  permissionIssues: string[]
  summary: Record<string, number>
  error?: string
  precision: number
  recall: number
  f1: number
}

function normalise(path: string): string {
  return path.replace(/\\/g, "/").toLowerCase()
}

function levenshteinRatio(a: string, b: string): number {
  if (!a || !b) return 0
  a = a.toLowerCase()
  b = b.toLowerCase()
  const n = a.length
  const m = b.length
  const dp: number[][] = Array.from({ length: n + 1 }, () => new Array(m + 1).fill(0))
  for (let i = 0; i <= n; i++) dp[i][0] = i
  for (let j = 0; j <= m; j++) dp[0][j] = j
  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= m; j++) {
      dp[i][j] = Math.min(
        dp[i - 1][j] + 1,
        dp[i][j - 1] + 1,
        dp[i - 1][j - 1] + (a[i - 1] === b[j - 1] ? 0 : 1),
      )
    }
  }
  return 1 - dp[n][m] / Math.max(n, m)
}

function extractClaimedFiles(prompt: string): string[] {
  const seen = new Set<string>()
  const files: string[] = []
  let m: RegExpExecArray | null
  FILE_REF_PATTERN.lastIndex = 0
  while ((m = FILE_REF_PATTERN.exec(prompt)) !== null) {
    const p = m[1]
    if (!seen.has(p) && !p.startsWith("http://") && !p.startsWith("https://") && !p.startsWith("www.")) {
      seen.add(p)
      files.push(p)
    }
  }
  return files
}

function scanActualFiles(directory: string): string[] {
  const base = resolve(directory)
  const files: string[] = []

  function walk(dir: string): void {
    let entries: string[]
    try {
      entries = readdirSync(dir)
    } catch {
      return
    }
    for (const entry of entries) {
      const abs = resolve(dir, entry)
      try {
        const s = statSync(abs)
        const rel = relative(base, abs).replace(/\\/g, "/")
        const parts = rel.split("/")
        if (parts.some((p) => EXTRA_IGNORE.has(p))) continue
        if (s.isFile()) {
          files.push(rel)
        } else if (s.isDirectory()) {
          walk(abs)
        }
      } catch {
        // skip unreadable
      }
    }
  }

  walk(base)
  return files.sort()
}

export function audit(prompt: string, directory = "."): AuditResult {
  const claimed = extractClaimedFiles(prompt)
  const actual = scanActualFiles(directory)

  const normalisedClaimed = claimed.map(normalise)
  const normalisedClaimedSet = new Set(normalisedClaimed)
  const normalisedActualSet = new Set(actual.map(normalise))

  const matched: string[] = []
  const missing: string[] = []
  for (let i = 0; i < claimed.length; i++) {
    if (normalisedActualSet.has(normalisedClaimed[i])) {
      matched.push(claimed[i])
    } else {
      missing.push(claimed[i])
    }
  }

  const nearMatches: NearMatch[] = []
  for (const m of missing) {
    for (const a of actual) {
      const r = levenshteinRatio(m, a)
      if (r >= 0.5 && r < 1.0) {
        nearMatches.push({ claimed: m, found: a, similarity: Math.round(r * 100) / 100 })
      }
    }
  }
  nearMatches.sort((a, b) => b.similarity - a.similarity)

  const nearActuals = new Set(nearMatches.map((nm) => normalise(nm.found)))
  const extra = actual.filter((a) => !normalisedClaimedSet.has(normalise(a)) && !nearActuals.has(normalise(a)))

  const claimedCount = claimed.length || 1
  const actualCount = actual.length || 1
  const precision = matched.length / claimedCount
  const recall = matched.length / actualCount
  const f1 = (2 * precision * recall) / Math.max(precision + recall, 0.001)

  const rPrecision = Math.round(precision * 1000) / 1000
  const rRecall = Math.round(recall * 1000) / 1000
  const rF1 = Math.round(f1 * 1000) / 1000

  return {
    prompt,
    claimedFiles: claimed,
    actualFiles: actual,
    missingFiles: missing,
    extraFiles: extra,
    matchedFiles: matched,
    nearMatches,
    permissionIssues: [],
    summary: {
      claimed: claimed.length,
      actual: actual.length,
      matched: matched.length,
      missing: missing.length,
      extra: extra.length,
      nearMatches: nearMatches.length,
      permissionIssues: 0,
      precision: rPrecision,
      recall: rRecall,
      f1: rF1,
    },
    precision: rPrecision,
    recall: rRecall,
    f1: rF1,
  }
}

export function renderAuditReport(result: AuditResult): string {
  const CHECK = "\u2713"
  const CROSS = "\u2717"
  const ARROW = "\u2192"
  const ELLIPSIS = "\u2026"
  const TILDE = "~"
  const sep = "=".repeat(56)

  const lines: string[] = []
  const total = result.summary.claimed
  const matched = result.summary.matched
  const missing = result.summary.missing
  const extra = result.summary.extra

  let verdict: string
  if (!missing && result.permissionIssues.length === 0) {
    verdict = `  ${CHECK}  All ${total} claimed files found.`
  } else {
    verdict = `  ${CROSS}  ${matched} of ${total} claimed files found  (${missing} missing${extra ? ", " + extra + " extra" : ""})`
  }

  lines.push(sep)
  lines.push("  ambiguity audit \u2014 file creation verification")
  lines.push(sep)
  lines.push("")
  lines.push(verdict)
  lines.push("")

  if (missing > 0) {
    lines.push("  Missing files:")
    for (const f of result.missingFiles.slice(0, 15)) {
      lines.push(`    ${CROSS}  ${f}`)
    }
    if (result.missingFiles.length > 15) {
      lines.push(`    ${ELLIPSIS} and ${result.missingFiles.length - 15} more`)
    }
    lines.push(`  ${ARROW}  Check for typos in the file path or name`)
    lines.push("")
  }

  if (result.nearMatches.length > 0) {
    lines.push("  Possible name mismatches:")
    for (const nm of result.nearMatches.slice(0, 5)) {
      lines.push(`    ${TILDE}  claimed: ${nm.claimed}  ${ARROW}  actual: ${nm.found}`)
    }
    lines.push(`  ${ARROW}  Rename the actual file to match the claimed name, or update the prompt`)
    lines.push("")
  }

  if (extra > 0) {
    lines.push(`  Extra files (${extra}):`)
    for (const f of result.extraFiles.slice(0, 15)) {
      lines.push(`    +  ${f}`)
    }
    if (result.extraFiles.length > 15) {
      lines.push(`    ${ELLIPSIS} and ${result.extraFiles.length - 15} more`)
    }
    lines.push(`  ${ARROW}  Verify these were intentionally created`)
    lines.push("")
  }

  if (result.permissionIssues.length > 0) {
    lines.push("  Permission issues:")
    for (const i of result.permissionIssues) {
      lines.push(`    !  ${i}`)
    }
    lines.push(`  ${ARROW}  Check file permissions and directory write access`)
    lines.push("")
  }

  lines.push(sep)
  if (missing || extra || result.permissionIssues.length) {
    lines.push(`  precision: ${result.precision.toFixed(3)}  recall: ${result.recall.toFixed(3)}  f1: ${result.f1.toFixed(3)}`)
  }
  lines.push(sep)
  return lines.join("\n")
}

File: src/ambiguity/audit.py
"""Audit claimed file creations against actual filesystem state."""

import os
import re
import sys
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

_USE_UNICODE = sys.stdout.encoding and sys.stdout.encoding.lower() in ("utf-8", "utf8")
CHECK = "\u2713" if _USE_UNICODE else "[OK]"
CROSS = "\u2717" if _USE_UNICODE else "[X]"
ARROW = "\u2192" if _USE_UNICODE else "->"
ELLIPSIS = "\u2026" if _USE_UNICODE else "..."
TILDE = "~"


FILE_REF_PATTERN = re.compile(
    r'(?:^|\s)(?:`)?'
    r'((?:[\w.-]+/)*[\w.-]+\.\w+)'
    r'(?:`)?(?=\s|$|[.,;!?])',
    re.MULTILINE,
)

EXTRA_IGNORE = {".git", "node_modules", "__pycache__", ".pytest_cache", ".ruff_cache", ".mypy_cache"}


@dataclass
class AuditResult:
    prompt: str
    claimed_files: list[str]
    actual_files: list[str]
    missing_files: list[str] = field(default_factory=list)
    extra_files: list[str] = field(default_factory=list)
    matched_files: list[str] = field(default_factory=list)
    near_matches: list[tuple[str, str, float]] = field(default_factory=list)
    permission_issues: list[str] = field(default_factory=list)
    summary: dict[str, Any] = field(default_factory=dict)
    error: str | None = None

    precision: float = 0.0
    recall: float = 0.0
    f1: float = 0.0


def _normalise(path: str) -> str:
    return path.replace("\\", "/").lower()


def _levenshtein_ratio(a: str, b: str) -> float:
    if not a or not b:
        return 0.0
    a, b = a.lower(), b.lower()
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(n + 1):
        dp[i][0] = i
    for j in range(m + 1):
        dp[0][j] = j
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            dp[i][j] = min(
                dp[i - 1][j] + 1,
                dp[i][j - 1] + 1,
                dp[i - 1][j - 1] + cost,
            )
    return 1.0 - dp[n][m] / max(n, m)


def _extract_claimed_files(prompt: str) -> list[str]:
    seen: set[str] = set()
    files: list[str] = []
    for m in FILE_REF_PATTERN.finditer(prompt):
        path = m.group(1)
        if path not in seen and not path.startswith(("http://", "https://", "www.")):
            seen.add(path)
            files.append(path)
    return files


def _scan_actual_files(directory: str) -> list[str]:
    base = Path(directory).resolve()
    files: list[str] = []
    for fp in base.rglob("*"):
        rel = fp.relative_to(base)
        parts = set(rel.parts)
        if EXTRA_IGNORE & parts:
            continue
        if fp.is_file():
            files.append(str(rel.as_posix()))
    return sorted(files)


def _check_permissions(claimed: list[str], directory: str) -> list[str]:
    issues: list[str] = []
    for path in claimed:
        fp = Path(directory) / path
        if fp.exists() or fp.parent.exists():
            if fp.exists() and not os.access(str(fp), os.R_OK):
                issues.append(f"{path}: no read permission")
            parent = fp.parent
            while parent != parent.parent:
                if parent.exists() and not os.access(str(parent), os.W_OK):
                    issues.append(f"{path}: parent {parent} not writable")
                    break
                parent = parent.parent
    return issues


def audit(prompt: str, directory: str = ".") -> AuditResult:
    claimed = _extract_claimed_files(prompt)
    actual = _scan_actual_files(directory)
    permission_issues = _check_permissions(claimed, directory)

    normalised_claimed_list = [_normalise(c) for c in claimed]
    normalised_claimed_set = set(normalised_claimed_list)
    normalised_actual_set = set(_normalise(a) for a in actual)

    matched: list[str] = []
    missing: list[str] = []
    for c, nc in zip(claimed, normalised_claimed_list):
        if nc in normalised_actual_set:
            matched.append(c)
        else:
            missing.append(c)

    near_matches: list[tuple[str, str, float]] = []
    for m in missing:
        for a in actual:
            ratio = _levenshtein_ratio(m, a)
            if ratio >= 0.5 and ratio < 1.0:
                near_matches.append((m, a, round(ratio, 2)))
    near_matches.sort(key=lambda x: -x[2])

    near_matched_actuals = set(_normalise(a) for _, a, _ in near_matches)
    extra: list[str] = [
        a for a in actual
        if _normalise(a) not in normalised_claimed_set
        and _normalise(a) not in near_matched_actuals
    ]

    claimed_count = len(claimed) or 1
    actual_count = len(actual) or 1
    matched_count = len(matched)
    precision = matched_count / max(claimed_count, 1)
    recall = matched_count / max(actual_count, 1)
    f1 = 2 * precision * recall / max(precision + recall, 0.001)

    summary = {
        "claimed": len(claimed),
        "actual": len(actual),
        "matched": len(matched),
        "missing": len(missing),
        "extra": len(extra),
        "near_matches": len(near_matches),
        "permission_issues": len(permission_issues),
        "precision": round(precision, 3),
        "recall": round(recall, 3),
        "f1": round(f1, 3),
    }
    result = AuditResult(
        prompt=prompt,
        claimed_files=claimed,
        actual_files=actual,
        missing_files=missing,
        extra_files=extra,
        matched_files=matched,
        near_matches=near_matches,
        permission_issues=permission_issues,
        summary=summary,
        precision=round(precision, 3),
        recall=round(recall, 3),
        f1=round(f1, 3),
    )
    return result


def render_audit_report(result: AuditResult) -> str:
    lines = []
    sep = "=" * 56

    total = result.summary["claimed"]
    matched = result.summary["matched"]
    missing = result.summary["missing"]
    extra = result.summary["extra"]

    if not missing and not result.permission_issues:
        verdict = f"  {CHECK}  All {total} claimed files found."
    else:
        detail = f"{matched} of {total}"
        verdict = f"  {CROSS}  {detail} claimed files found  ({missing} missing{', ' + str(extra) + ' extra' if extra else ''})"

    lines.append(sep)
    lines.append("  ambiguity audit \u2014 file creation verification")
    lines.append(sep)
    lines.append("")
    lines.append(verdict)
    lines.append("")

    if missing:
        lines.append("  Missing files:")
        for f in result.missing_files[:15]:
            lines.append(f"    {CROSS}  {f}")
        if len(result.missing_files) > 15:
            lines.append(f"    {ELLIPSIS} and {len(result.missing_files) - 15} more")
        lines.append(f"  {ARROW}  Check for typos in the file path or name")
        lines.append("")

    if result.near_matches:
        lines.append("  Possible name mismatches:")
        for claimed_f, actual_f, ratio in result.near_matches[:5]:
            lines.append(f"    {TILDE}  claimed: {claimed_f}  {ARROW}  actual: {actual_f}")
        lines.append(f"  {ARROW}  Rename the actual file to match the claimed name, or update the prompt")
        lines.append("")

    if extra:
        lines.append(f"  Extra files ({extra}):")
        for f in result.extra_files[:15]:
            lines.append(f"    +  {f}")
        if len(result.extra_files) > 15:
            lines.append(f"    {ELLIPSIS} and {len(result.extra_files) - 15} more")
        lines.append(f"  {ARROW}  Verify these were intentionally created")
        lines.append("")

    if result.permission_issues:
        lines.append("  Permission issues:")
        for i in result.permission_issues:
            lines.append(f"    !  {i}")
        lines.append(f"  {ARROW}  Check file permissions and directory write access")
        lines.append("")

    lines.append(sep)
    if missing or extra or result.permission_issues:
        lines.append(f"  precision: {result.precision:.3f}  recall: {result.recall:.3f}  f1: {result.f1:.3f}")
    lines.append(sep)
    return "\n".join(lines)


def _compute_summary(result: AuditResult) -> dict[str, Any]:
    return {
        "claimed": len(result.claimed_files),
        "actual": len(result.actual_files),
        "matched": len(result.matched_files),
        "missing": len(result.missing_files),
        "extra": len(result.extra_files),
        "near_matches": len(result.near_matches),
        "permission_issues": len(result.permission_issues),
        "precision": result.precision,
        "recall": result.recall,
        "f1": result.f1,
    }


def render_audit_json(result: AuditResult) -> dict[str, Any]:
    summary = _compute_summary(result)
    return {
        "command": "audit",
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "prompt": result.prompt,
        "claimed_files": result.claimed_files,
        "actual_files": result.actual_files,
        "missing_files": result.missing_files,
        "extra_files": result.extra_files,
        "matched_files": result.matched_files,
        "near_matches": [(a, b, r) for a, b, r in result.near_matches],
        "permission_issues": result.permission_issues,
        "summary": summary,
    }
